### PR TITLE
[Small Feature] Region Growing on Spheres and Cylinders

### PR DESCRIPTION
**Pull request by sgiraudot**
## Rationale

The _Shape Detection_ package provides 2 algorithms: RANSAC and Region Growing. Experimentally, Region Growing is much more reliable but can only detect planes (or lines in 2D) while RANSAC can also detect spheres, cylinders, tori and cones.

This small feature adds the following shapes to Region Growing:
- circles in 2D
- spheres in 3D
- cylinders in 3D

Tori and cones seem quite anecdotal and are ignored (at least until it becomes clear that they would indeed be used if present).

So far, one limitation is that Region Growing only allows to detect one type at a time, so contrary to RANSAC, it won't be possible to search for planes, spheres and cylinders at the same time in a scene.

## Summary of API changes

Three new classes are added:
- `CGAL::Shape_detection::Point_set::Least_squares_circle_fit_region`
- `CGAL::Shape_detection::Point_set::Least_squares_sphere_fit_region`
- `CGAL::Shape_detection::Point_set::Least_squares_cylinder_fit_region`

Each class comes with an additional example.

## License and copyright ownership

(no change)

## CHANGES.md

TODO

## Submission

* [__User Manual__](https://cgal.github.io/5321/v1/Shape_detection/index.html#title10) (modified the introduction to mention the new classes + 3 new examples)
* [__Reference Manual__](https://cgal.github.io/5321/v1/Shape_detection/group__PkgShapeDetectionRGOnPoints.html) (3 new classes)

## Status

* Developed and locally tested (GNU/Linux)
* [Small Feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Region_Growing_on_Spheres_and_Cylinders)  -- **pre-approved** 2021/04/15

### CODE REVIEW

File: Shape_detection/examples/Shape_detection/region_growing_cylinders_on_point_set_3.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+using Neighbor_query = Shape_detection::K_neighbor_query
+  <Kernel, Point_set, Point_map>;
+using Region_type = Shape_detection::Least_squares_cylinder_fit_region
+  <Kernel, Point_set, Point_map, Normal_map>;
+using Region_growing = CGAL::Shape_detection::Region_growing
+  <Point_set, Neighbor_query, Region_type>;
+
+int main (int argc, char** argv)
+{
+  std::ifstream ifile (argc > 1 ? argv[1] : "data/cube.pwn");
```
Comment (danston) [inline Shape_detection/examples/Shape_detection/region_growing_cylinders_on_point_set_3.cpp:34]: Do not you need to load here something else than the cube data set? As far as I remember, it contains not only cylinders so we run into an issue of detecting multiple objects with RG, which does not sell well the algorithm.

File: Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_circle_fit_region.h [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    circle. %Default is 1.
+
+    \param angle_threshold the maximum accepted angle in degrees
+    between the normal of a point and the radius of a circle. %Default
+    is 25 degrees.
+
+    \param min_region_size the minimum number of 2D points a region
+    must have. %Default is 3.
+
+    \param minimum_radius is the radius below which an estimated
```
Comment (danston) [inline Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_circle_fit_region.h:134]: I would remove `is` to be consistent with other parameters like `angle_threshold` e.g. The same for `maximum_radius` and also for other classes with spheres and cylinders.

File: Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_circle_fit_region.h [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    circle is considered as invalid and discarded. %Default is 0 (no
+    limit).
+
+    \param maximum_radius is the radius above which an estimated
```
Comment (danston) [inline Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_circle_fit_region.h:138]: Remove `is`.

File: Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_cylinder_fit_region.h [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+    cylinder. %Default is 1.
+
+    \param angle_threshold the maximum accepted angle in degrees
+    between the normal of a point and the radius of a cylinder. %Default
+    is 25 degrees.
+
+    \param min_region_size the minimum number of 3D points a region
+    must have. %Default is 3.
+
+    \param minimum_radius is the radius below which an estimated
```
Comment (danston) [inline Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_cylinder_fit_region.h:135]: Remove `is`.

File: Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_cylinder_fit_region.h [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    cylinder is considered as invalid and discarded. %Default is 0 (no
+    limit).
+
+    \param maximum_radius is the radius above which an estimated
```
Comment (danston) [inline Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_cylinder_fit_region.h:139]: Remove `is`.

File: Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_sphere_fit_region.h [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+    sphere. %Default is 1.
+
+    \param angle_threshold the maximum accepted angle in degrees
+    between the normal of a point and the radius of a sphere. %Default
+    is 25 degrees.
+
+    \param min_region_size the minimum number of 3D points a region
+    must have. %Default is 3.
+
+    \param minimum_radius is the radius below which an estimated
```
Comment (danston) [inline Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_sphere_fit_region.h:139]: Remove `is`.

File: Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_sphere_fit_region.h [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    sphere is considered as invalid and discarded. %Default is 0 (no
+    limit).
+
+    \param maximum_radius is the radius above which an estimated
```
Comment (danston) [inline Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_point_set/Least_squares_sphere_fit_region.h:143]: Remove `is`.

**Review by danston — CHANGES_REQUESTED** [review]
Hi. Great work. I really like this feature. There are a few small things to fix but otherwise, all is good. If you have any objections, please let me know. Finally, I have a few questions: 

1. What about changing the picture for the 2D example and make something that shows all types of detected shapes, the same for 3D points? 

2. I am wondering if it is possible to add the pre-sorting of input points based on the quality of the circle/sphere/cylinder fit as we have done it for 2D lines and 3D planes? Will it improve the results? Does it make sense?

3. What about spheres and cylinders on a polygon mesh? Is it easy to add and does it make sense?

Thanks.

**Comment by sgiraudot** [comment]
Thanks for the review, I just fixed what was suggested!

**Comment by maxGimeno** [comment]
There are conflicts with master

**Comment by sgiraudot** [comment]
Fixed.

**Comment by sgiraudot** [comment]
@danston I just realized I integrated your review's remarks but forgot to answer your first message.

>     1. What about changing the picture for the 2D example and make something that shows all types of detected shapes, the same for 3D points?

I am not sure this really makes sense, as we probably don't want to change the figure everytime we add a new type of shape. I think saying that we also support spheres/cylinders/etc. is pretty self-explanatory without needing to actually show it.

>     2. I am wondering if it is possible to add the pre-sorting of input points based on the quality of the circle/sphere/cylinder fit as we have done it for 2D lines and 3D planes? Will it improve the results? Does it make sense?

This is a good question. I don't know but it's most probably possible: I'll check this out and see if I can add it.

>     3. What about spheres and cylinders on a polygon mesh? Is it easy to add and does it make sense?

It would definitely make sense. It's probably a bit harder to do though, as it would require an integral formulation on the surfaces, which I'm not sure I know how to do. I'll have a look too, maybe I can take inspiration from the PCA package.

**Review by danston — APPROVED** [review]

**Comment by sgiraudot** [comment]
@danston

I just pushed some additional features: sorting classes for spheres, cylinders and circles. I added tests and documentation for these. It's pretty much copy pasted from your other sorting classes, but if you can have a quick look to check that I did not mess anything up, it's cool!

About detecting cylinders and spheres on polygon meshes: I had a look, and I don't think I can do it. We would need to have an integral formulation similar to what's done on PCA, but the formulation for spheres is quite different from the one on planes, and I can't figure out how to do it (for cylinders, it's even more complicated as I don't have an integral formulation for planes either and use a RANSAC-type of detection instead). If someone finds a way to do this properly in the future, it'd still would be a nice feature.

Another idea for future implementations: there could be an additional neighborhood structure for polygon meshes that uses edges that connect two vertices. That way, we could do region growing on _the vertices_ of a mesh, which means we could use the point set's region types (including spheres and cylinders) while taking advantage of the well-defined neighborhood of the edge graph of the mesh. It wouldn't be as useful as the face-based detection, but it would be a way to have more shapes detected in polygon meshes.

**Review by danston — APPROVED** [review]
The changes look good!

**Comment by danston** [comment]
Hi @sgiraudot, 
- I checked your changes. They look good and correct!
- Since getting new shapes on polygon meshes seems to be complicated, I do not think we should bother adding them now though it could be a great task for the next GSoC in case I find a good student for that.
- Your last idea is very nice. Actually, in my new PR, I have something like that by extracting a set of edges from the polygon mesh and running region growing on this set. I will see later if I can find a good way to do what you proposed by using a filtered face graph e.g.

**Comment by danston** [comment]
There are some errors in the build. I think they come from missing includes.

**Comment by maxGimeno** [comment]
This kind of errors : 
```
/home/runner/work/cgal/cgal/Shape_detection/include/CGAL/Shape_detection/Region_growing/internal/fitting.h:186:35: error: ‘property_traits’ in namespace ‘boost’ does not name a template type
2021-05-04T06:56:06.0595342Z   186 |   using Point_3 = typename boost::property_traits<PointMap>::value_type;
```

**Comment by danston** [comment]
Substituted by #5808.

**Pull request closed without merging**